include/ginkgo/core/matrix/csr.hpp

```diff
@@ -314,13 +356,79 @@ class Csr : public EnableLinOp<Csr<ValueType, IndexType>>,
 
         int64_t clac_size(const int64_t nnz)
         {
-            return std::make_shared<load_balance>(nwarps_)->clac_size(nnz);
+            return std::make_shared<load_balance>(nwarps_, warp_size_,
+                                                  cuda_strategy_)
+                ->clac_size(nnz);
         }
 
     private:
         int64_t nwarps_;
+        int warp_size_;
+        bool cuda_strategy_;
     };
 
+    void convert_to(Csr<ValueType, IndexType> *result) const override
+    {
+        auto executor = result->get_executor();
+        bool same_executor = this->get_executor() == executor;
+        auto strategy = this->get_strategy();
+        EnableLinOp<Csr>::convert_to(result);
+        if (!same_executor) {
+            // When strategy is load_balance or automatical, rebuild the
+            // strategy according to executor's property.
+            if (std::dynamic_pointer_cast<load_balance>(strategy)) {
+                if (auto exec = std::dynamic_pointer_cast<const HipExecutor>(
+                        executor)) {
+                    result->set_strategy(std::make_shared<load_balance>(exec));
+                } else if (auto exec =
+                               std::dynamic_pointer_cast<const CudaExecutor>(
+                                   executor)) {
+                    result->set_strategy(std::make_shared<load_balance>(exec));
+                }
+            } else if (std::dynamic_pointer_cast<automatical>(strategy)) {
+                if (auto exec = std::dynamic_pointer_cast<const HipExecutor>(
+                        executor)) {
+                    result->set_strategy(std::make_shared<automatical>(exec));
+                } else if (auto exec =
+                               std::dynamic_pointer_cast<const CudaExecutor>(
+                                   executor)) {
+                    result->set_strategy(std::make_shared<automatical>(exec));
+                }
+            }
```

Comment:
I would put this into a helper a helper
```c++
void strategy_rebuild_helper(Csr<ValueType, IndexType> *result) { ... }
```

Then the content of these functions become:
```c++
EnableLinOp<Csr>::convert_to(result);
strategy_rebuild_helper(result);
```